File: 数据结构篇/二叉树/遍历/前序遍历/144. 二叉树的前序遍历.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def preorderTraversal(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        
        
        # 非递归
        stack=[]
        res=[]
        
        stack.append(root)
        
        while len(stack)>0:
            cur=stack.pop()
            if(cur):
                res.append(cur.val)
            else:
                continue
            stack.append(cur.right)
            stack.append(cur.left)
        return res
        
    #     递归    
    #     self.res=[]
    #     if not root:
    #         return []
    #     self.DFS(root)
    #     return self.res
        
    # def DFS(self,root):
    #     if not root: 
    #         return
    #     self.res.append(root.val)
    #     self.DFS(root.left)
    #     self.DFS(root.right)
        
        




        

File: 数据结构篇/二叉树/遍历/前序遍历/144.二叉树的前序遍历.cpp


#include <vector>
#include <queue>
#include <iostream>
using namespace std;

struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
    /* data */
};

class Solution
{
public:
    void travelsal(TreeNode *cur, vector<int> &vec)
    {
        if (cur == NULL)
            return;
        vec.push_back(cur->val);
        travelsal(cur->left, vec);
        travelsal(cur->right, vec);
    }

    vector<int> preoderTraversal(TreeNode *root){
        vector<int> vec;
        travelsal(root,vec);
        return vec;
    }

private:
};

int main()
{
    TreeNode *root = new TreeNode(1);
    root->left = new TreeNode(2);
    root->right = new TreeNode(3);
    root->left->left = new TreeNode(4);
    root->left->right = new TreeNode(5);

    Solution solution;
    vector<int> vec=solution.preoderTraversal(root);
    for (auto i : vec)
    {
        std::cout << i << std::endl;
    }
}